torchmetrics/metric.py

```diff
@@ -415,25 +417,68 @@ def state_dict(
         destination: Dict[str, Any] = None,
         prefix: str = "",
         keep_vars: bool = False,
+        should_sync: Optional[bool] = None
     ) -> Optional[Dict[str, Any]]:
         destination = super().state_dict(destination=destination, prefix=prefix, keep_vars=keep_vars)
         # Register metric states to be part of the state_dict
-        with self.sync_context(dist_sync_fn=self.dist_sync_fn):
+
+        # whether the metric should be synced.
+        should_sync = should_sync if should_sync is not None else self.dist_sync_state_dict
+        is_persistent = False
+
+        with self.sync_context(dist_sync_fn=self.dist_sync_fn, should_sync=should_sync):
+            device = None
             for key in self._defaults:
+
+                # used to track if any defaults is persistent
+                is_persistent |= self._persistent[key]
+
                 if not self._persistent[key]:
                     continue
-                current_val = getattr(self, key)
+
+                if should_sync and not self.is_global_zero:
+                    current_val = self._defaults[key]
+                else:
+                    current_val = getattr(self, key)
+
                 if not keep_vars:
                     if isinstance(current_val, Tensor):
                         current_val = current_val.detach()
+                        if device is None:
+                            device = current_val.device
                     elif isinstance(current_val, list):
                         current_val = [cur_v.detach() if isinstance(cur_v, Tensor) else cur_v for cur_v in current_val]
+                        for v in current_val:
+                            if device is None:
+                                device = v.device
+
                 # the tensors will be synced across processes so deepcopy to drop the references
                 destination[prefix + key] = deepcopy(current_val)  # type: ignore
+
+            if is_persistent:
+                destination[prefix + "has_synced"] = torch.tensor(should_sync, device=device)  # type: ignore
+                destination[prefix + "rank"] = torch.tensor(self.current_rank, device=device)  # type: ignore
```

Comment:
Why do we need to save them and why as tensors?

Reply:
For DDP. If the LightningModule is wrapped into a DDP wrapped, parameters will be broadcasted for check.
Therefore they need to be tensors.
I plan to add a way to bypass this, but this need to be added to Lightning first.